Python/Module3_IntroducingNumpy/Broadcasting.md

```diff
@@ -620,34 +638,106 @@ Having accounted for all three terms, we can finally compute the Euclidean dista
 
-        Returns
-        -------
-        numpy.ndarray, shape=(M, N)
-            The Euclidean distance between each pair of
-            rows between `x` and `y`."""
-    dists = -2 * np.matmul(x, y.T)
-    dists +=  np.sum(x**2, axis=1)[:, np.newaxis]
-    dists += np.sum(y**2, axis=1)
-    return  np.sqrt(dists)
+```python
+# The square-distances should be exactly zero, but they
+# will instead be (*very* small) negative numbers!
+>>> sqr_dists = -2 * np.matmul(x, x.T)  # `x_x_prod`
+>>> sqr_dists +=  np.sum(x**2, axis=1)[:, np.newaxis]
+>>> sqr_dists += np.sum(x**2, axis=1)
+>>> sqr_dists
+array([[-2.842170943040401e-14, -2.842170943040401e-14],
+       [-2.842170943040401e-14, -2.842170943040401e-14]])
 ```
 
+These values are *very* close to being zero, so it is not as if the magnitude of our result is wildly off, but the critical issue here is that the quirks of floating point arithmetic produced (very small) negative numbers.
+We are going to be in for a rude awakening when we take the square-root of these values to get our final distances:
 
+```python
+# Taking the square-root of negative floats will produce NaNs
+>>> np.sqrt(sqr_dists)
+array([[nan, nan],
+       [nan, nan]])
+```
+
+`nan` stands for "Not a Number", which is to say that the square-root of a negative number cannot produce a real-valued float (the result will be an [imaginary number](https://www.pythonlikeyoumeanit.com/Module2_EssentialsOfPython/Basic_Objects.html#Complex-Numbers), but NumPy will not swap out real numbers for imaginary ones unless we explicitly permit it to).
+
+The solution to this problem comes from the realization that this occurrence is truly an edge-case, and that it will only manifest when the result *ought* to have been zero and indeed is very close to zero.
+Thus we can [clip](https://numpy.org/doc/stable/reference/generated/numpy.clip.html) `sqr_dists` such that any value that falls below zero gets set to zero and all other values will be left alone:
+
+```python
+# "Clipping" all negative entries so that they are set to 0 
+>>> np.sqrt(np.clip(sqr_dists, a_min=0, a_max=None))
+array([[0., 0.],
+       [0., 0.]])
+```
+<!-- #endregion -->
+
+<!-- #region -->
+#### The Final Answer, At Last!
+
+In total, we have successfully used vectorization to compute the all pairs of distances, while only requiring an array of shape-$(M, N)$ to do so! This is the memory-efficient, vectorized form – the stuff that dreams are made of.
+Let's write the function that performs this computation in full.  
+
+```python
+def pairwise_dists(x, y):
+    """ Computing pairwise distances using memory-efficient 
+    vectorization.
+
+    Parameters
+    ----------
+    x : numpy.ndarray, shape=(M, D)
+    y : numpy.ndarray, shape=(N, D)
+
+    Returns
+    -------
+    numpy.ndarray, shape=(M, N)
+        The Euclidean distance between each pair of
+        rows between `x` and `y`."""
```

Comment:
do you want to leave the end-quote on the same line or a newline? not sure what the consistent thing is off the top of my head

Reply:
I think same line is what `black` does. So I'll go with it.